### ISSUE TITLE
Application startup exception

### ISSUE THREAD

**Issue by junpark12** [Microsoft/ApplicationInsights-Kubernetes#120]
Hi. When i start my asp.net 2.0 core app using this document, i received the following exception.

info: Microsoft.ApplicationInsights.Kubernetes.KubernetesModule[0]
      ApplicationInsights.Kubernetes.Version:ai-k8s:1.0.0.7
Application startup exception: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion()
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at Autofac.Extensions.DependencyInjection.AutofacRegistration.Register(ContainerBuilder builder, IEnumerable`1 descriptors, Object lifetimeScopeTagForSingletons)
   at ContainerTest.Startup.ConfigureServices(IServiceCollection services) in /src/ContainerTest/Startup.cs:line 80

If i comment out "service.EnableKubernetes()", i do not receive any errors.

            services.Configure<AppSettings>(Configuration);
            services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
            services.AddTransient<ITodoService, TodoService>();
            services.AddSingleton<IHttpClient, StandardHttpClient>();
            //services.EnableKubernetes();
            var container = new ContainerBuilder();
            container.Populate(services);

            return new AutofacServiceProvider(container.Build());

Is this know issue?
Thanks.

**Comment by xiaomi7732** [comment]
@junpark12 Thanks for reporting this issue. This is not known. Please give me some time to investigate the issue.
I am not familiar with AutofacServiceProvider. Is it a IoC provider that provides all service instances from the asp.net service collection?
Thanks!

**Comment by xiaomi7732** [comment]
Oh, as a workaround, could you please try:
```csharp
using System.Linq;
// ...
        services.EnableKubernetes();
        var container = new ContainerBuilder();
        container.Populate(services.ToList());
```

**Comment by junpark12** [comment]
@xiaomi7732  it's working like a charm. Thank you for the help :)

**Comment by xiaomi7732** [comment]
@junpark12 thanks for the try. 👍 
That is just a workaround. I will spend sometime to fix the real issue. I hope you don't mind me reopen this issue for tracking :-)

**Comment by junpark12** [comment]
@xiaomi7732  sure. If there are any update , let me know. Thanks.

**Comment by xiaomi7732** [comment]
The issue is here:
```csharp
            // Dispatch this on a differnet thread to avoid blocking the main thread.
            // Mainly used with K8s Readness Probe enabled, where communicating with Server will temperory be blocked.
            // TODO: Instead of query the server on the start, we should depend on watch services to provide dynamic realtime data.
            Task.Run(() =>
            {
                KubernetesModule.EnableKubernetes(services, timeout, kubernetesServiceCollectionBuilder);
            });

            return services;
```
Basically, the kubernetesServiceCollectionBuilder is building and injecting the services into the services; However, since it is dispatched onto a different thread, the return of the services collection could/and will happen before all the services are injected. That is a bad practice.

In your code, if I guess correct, when you all `container.Populate(services);`, it will enumerate all the services in the service collection. At this time, the kubernetesServiceCollectionBuilder is still injecting new services, that will break the enumerator, leading to the exception.

* Why the workaround works?
Because when `ToList()` is called on an `IEnumerable`, it will realize the list and return it. Although the kubernetesServiceCollectionBuilder is still appending service to the service collection, the service list won't change anymore - thus you don't see the exception. But I bet you will not being able to realize any instances provided by kuberentesServiceCollectionBuilder.
That workaround is okay in this case because all application insights kubernetes related services (classes or interfaces) are internal that you either way will not access them through the service provider. So, your IoC container still includes all the types you registered.

* How is it fixed?
Interestingly, I happen to rewrite the model of injecting the services into the collection in to synchronized code in #119 :
```csharp
// ...

             serviceCollection.AddLogging();

            _logger.LogInformation(Invariant($"ApplicationInsights.Kubernetes.Version:{SDKVersionUtils.Instance.CurrentSDKVersion}"));
            try
            {
                serviceCollection = BuildK8sServiceCollection(serviceCollection, timeout.Value, kubernetesServiceCollectionBuilder);
            }
            catch (Exception ex)
            {
                _logger.LogError("Failed to fetch ApplicaitonInsights.Kubernetes' info. Details " + ex.ToString());
            }
            return serviceCollection;
```
And I verified the issue will be gone with a private package. I think it will released in beta8.

**Comment by xiaomi7732** [comment]
Released in 1.0.0-beta 8
